### PR TITLE
Export valid CMake CONFIG files for MaterialX

### LINKED ISSUE #642: Export CMake config file
Move from an obsolete MODULE type, to a more powerful CONFIG type of CMake package.

### PR DESCRIPTION
**Pull request by JGamache-autodesk**
Fixes #642

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -207,6 +207,24 @@ if(${CMAKE_VERSION} VERSION_GREATER "3.6.2")
...
+    ${CMAKE_BINARY_DIR}/cmake/MaterialXConfigVersion.cmake
+    VERSION ${CMAKE_VERSION}
+    COMPATIBILITY AnyNewerVersion
+)
+
+# Package for using the developper build
+export(
+    EXPORT MaterialX
+    FILE ${CMAKE_BINARY_DIR}/cmake/MaterialXConfig.cmake
+    )
```
Comment (JGamache-autodesk) [inline CMakeLists.txt]: Gathers all the info for targets with an EXPORT tag (see other CMakeLists.txt) and write it down as a CMake package.

File: source/MaterialXContrib/CMakeLists.txt [hunk 2]
```diff
@@ -1,8 +1,3 @@
-include_directories(    
-    ${EXTERNAL_INCLUDE_DIRS}
-    ${CMAKE_CURRENT_SOURCE_DIR}/../
-)
```
Comment (JGamache-autodesk) [inline source/MaterialXContrib/CMakeLists.txt]: Better to use "target_include_directories" as this allows information about header files to be written into the CMake package file.

File: source/MaterialXContrib/CMakeLists.txt [hunk 3]
```diff
@@ -31,10 +26,18 @@ target_link_libraries(
     MaterialXRender
     ${CMAKE_DL_LIBS})
 
-install(TARGETS MaterialXContrib
-    DESTINATION ${CMAKE_INSTALL_PREFIX}/lib/
+target_include_directories(MaterialXContrib
+    PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../>
+        $<INSTALL_INTERFACE:include>
```
Comment (JGamache-autodesk) [inline source/MaterialXContrib/CMakeLists.txt]: Using the MaterialXContrib target will now imply we need these include directories when used with a packaged MaterialX.

File: source/MaterialXContrib/CMakeLists.txt [hunk 4]
```diff
@@ -31,10 +26,18 @@ target_link_libraries(
...
+    PRIVATE
+        ${EXTERNAL_INCLUDE_DIRS}
 )
 
+install(TARGETS MaterialXContrib
+    EXPORT MaterialX
```
Comment (JGamache-autodesk) [inline source/MaterialXContrib/CMakeLists.txt]: Specify we want to export the build information for this target.

**Comment by bernardkwok** [comment]
@JGamache-autodesk : FindMaterialX.cmake was added by @ashwinbhat and could be used for the interop fork. Please check that removing this and the other changes don't break anything for him. Thanks.

**Comment by ashwinbhat** [comment]
This looks much better. I'm not a cmake expert. This is how I'm using this

https://git.autodesk.com/GFX/interop/blob/da46aa234340e6be401692c681787e5370de3bea/CMakeLists.txt#L8-L17

https://git.autodesk.com/GFX/interop/blob/bhata/mtlxGen/source/materialxassetutility/CMakeLists.txt#L1-L35

**Comment by JGamache-autodesk** [comment]
Changes in the first file:
`    set(MaterialX_DIR "${MATERIALX_LOCATION}/cmake")
    find_package(MaterialX REQUIRED)`
Changes in second file:
   - ${MATERIALX_INCLUDE_DIRS} no longer needed (targets know their include paths)
   - ${MATERIALX_CORE_LIBS} and ${MATERIALX_GENERATOR_LIBS} no longer exist. You can now directly add MaterialXGenShader in the target_link_libraries call and the includes and dependendent libraries will be handled automatically.

File: CMakeLists.txt [hunk 5]
```diff
@@ -207,6 +207,24 @@ if(${CMAKE_VERSION} VERSION_GREATER "3.6.2")
...
-install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/cmake"
-    DESTINATION "${CMAKE_INSTALL_PREFIX}" MESSAGE_NEVER)
+include(CMakePackageConfigHelpers)
+write_basic_package_version_file(
+    ${CMAKE_BINARY_DIR}/cmake/MaterialXConfigVersion.cmake
+    VERSION ${CMAKE_VERSION}
+    COMPATIBILITY AnyNewerVersion
+)
+
+# Package for using the developper build
```
Comment (feldstj) [inline CMakeLists.txt]: minor: typo developer

**Comment by ashwinbhat** [comment]
Pasting slack conv with @JGamache-autodesk for reference: 

The lib and include, are part of the dependencies of the MaterialX target libraries, so if you want to use MaterialXGenShader, you simply put that in the target_dependent_libs and CMake will automatically add the correct paths to the libs and include files as found in the config file.
It becomes a one line solution instead of having to remember adding the include path to the includes, the lib path to the lib dirs, and so on.
BTW, the install now looks like:
./materialx
./materialx/RelWithDebInfo
./materialx/RelWithDebInfo/bin
./materialx/RelWithDebInfo/cmake
**./materialx/RelWithDebInfo/cmake/MaterialXConfig-relwithdebinfo.cmake
./materialx/RelWithDebInfo/cmake/MaterialXConfig.cmake
./materialx/RelWithDebInfo/cmake/MaterialXConfigVersion.cmake**
./materialx/RelWithDebInfo/include
./materialx/RelWithDebInfo/lib
./materialx/RelWithDebInfo/libraries
./materialx/RelWithDebInfo/resources

**Pull request merged**